### PR TITLE
Core/SAI: Implemented new action SMART_ACTION_COMPLETE_QUEST

### PR DESCRIPTION
**Pull request by meji46**
**Changes proposed:**

-  Implemented new action SMART_ACTION_COMPLETE_QUEST. Only for quests using QUEST_FLAGS_COMPLETION_EVENT, QUEST_FLAGS_COMPLETION_AREA_TRIGGER or QUEST_FLAGS_TRACKING_EVENT (regular quests with objectives can't be completed using this action).
-  Deprecate SMART_ACTION_CALL_AREAEXPLOREDOREVENTHAPPENS and SMART_ACTION_CALL_GROUPEVENTHAPPENS. We can handle the same with the new action. Special cases can be done by using the appropriate target_type
- Update Spell::EffectQuestComplete
- Drop case for deprecated event_type SMART_EVENT_IS_BEHIND_TARGET

**Issues addressed:**
None

**Tests performed:**
Builds, tested in-game

**Known issues and TODO list:** (add/remove lines as needed)
None

### CODE REVIEW

File: src/server/game/AI/SmartScripts/SmartScriptMgr.cpp [hunk 1]
```diff
@@ -340,7 +340,6 @@ void SmartAIMgr::LoadSmartAIFromDB()
             }
             break;
         case SMART_EVENT_VICTIM_CASTING:
-        case SMART_EVENT_IS_BEHIND_TARGET:
```
Comment (jackpoz) [inline src/server/game/AI/SmartScripts/SmartScriptMgr.cpp:343]: Is this change related to the PR ?

Comment (meji46) [reply src/server/game/AI/SmartScripts/SmartScriptMgr.cpp:343]: Sneaky change for an already deprecated event

Comment (Faq) [reply src/server/game/AI/SmartScripts/SmartScriptMgr.cpp:343]: that should be mentioned under **Changes proposed:**

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 2]
```diff
@@ -2637,6 +2595,34 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
...
+                break;
+
+            for (WorldObject* target : targets)
+            {
+                Player* player = Object::ToPlayer(target);
+                if (!player)
+                    continue;
+
+                QuestStatus questStatus = player->GetQuestStatus(questId);
+                if (questStatus == QUEST_STATUS_REWARDED)
```
Comment (mdX7) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:2612]: should we check completed status aswell here?

Comment (meji46) [reply src/server/game/AI/SmartScripts/SmartScript.cpp:2612]: we're already checking for status INCOMPLETE here https://github.com/TrinityCore/TrinityCore/pull/30048/files#diff-6d3483b74df32d157a6f8f227dacec1496d15fadfa8da3300e87fe5f6166e213R2617 and the tracking quests should only change status from NONE to REWARDED (COMPLETE doesn't make sense for those quests)

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 3]
```diff
@@ -2637,6 +2595,34 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
...
+                    continue;
+
+                if (quest->HasFlag(QUEST_FLAGS_COMPLETION_EVENT) || quest->HasFlag(QUEST_FLAGS_COMPLETION_AREA_TRIGGER))
+                {
+                    if (questStatus == QUEST_STATUS_INCOMPLETE)
+                        player->AreaExploredOrEventHappens(questId);
+                }
+                else if (quest->HasFlag(QUEST_FLAGS_TRACKING_EVENT))  // Check if the quest is used as a serverside flag
+                        player->CompleteQuest(questId);
```
Comment (Shauren) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:2621]: bad indent

**Review by Shauren — CHANGES_REQUESTED** [review]

**Comment by Shauren** [comment]
Which quest scripts relied on that vehicle special case code to function?

**Comment by meji46** [comment]
I didn't see any cases relied on this code (not 100% sure that I didn't miss a specific case)

**Review by mdX7 — APPROVED** [review]
